### ISSUE TITLE
useDraggable with absolute positioning

### ISSUE THREAD

**Issue by calebwaldner** [vueuse/vueuse#1288]
I'd like to use useDraggable within a container component via relative/absolute positioning. useDraggable works well with a fixed position, but when I use an absolute position I get unexpected behavior. It seems the x/y coords do not respect the relative container and instead continue to calculate the fixed position x/y values.

Maybe I'm using the component wrong, but I spent a bit of time on this with no success.

in a nutshell, I want the draggable component to be absolute to the relative container and show coords relative to it's position within the container, not it's position on the screen.

Hope this makes sense, thanks for any help.

**Comment by alexrafuse** [comment]
Did you make any progress with this? I'm hitting the same issues with relative containers

**Comment by thomasaull** [comment]
In my app I used `portal-vue` together with `useDraggable` for exactly this problem and it works rather well :)

**Comment by stefnotch** [comment]
I did take a stab at this, and ended up running into thing to watch out for when implementing this for `position: absolute`:
Scrolling the page changes the "absolute" position of the mouse. So at the beginning, one would have to note down the current scrolled distance (and not forget about correctly handling nested scrollbars), and account for this while moving the element.
The simpler option of using `pageX` and `pageY` doesn't work, since scrolling doesn't emit any pointer events.



It might be easier to always use `position: fixed` during dragging, like:
- When the user starts dragging, use `getBoundingClientRect` to get the viewport position of the element. Then, change the element to `position: fixed`
- Dragging code works the way you'd expect. Crucially, scrolling also just works.
- When the user finishes dragging, one has to change the element back (back to `position: absolute`), *and calculate the correct position.* I'm not sure how to best go about that last part.

**Comment by stephenjason89** [comment]
I ended up using https://github.com/kirillmurashov/vue-drag-resize
I hope someone can fix this to be used with relative/absolute positioning.

**Comment by cmcnicholas** [comment]
same issue here, creating relative drag containers is causing this to be unusable. I imagine most cases need relative drag positioning it's rare in applications to just have completely floating panels unless they are just modals.

**Comment by cmcnicholas** [comment]
I have however found a workaround using just vueuse.

```ts
// do this on your container element, if in another vue component save state to store etc.
const { top: boundsTop, left: boundsLeft } = useElementBounding(container);

// whatever the initial values are
const initialLeft = ref(200);
const initialTop = ref(400);

// setup use draggable, collect the x,y raw not style
const { x: dragX, y: dragY, isDragging } = useDraggable(element);

// generate style manually adjusting left/top
const style = computed(() => {
  const adjustedLeft = unref(isDragging) ? unref(dragX) - unref(boundsLeft) : unref(initialLeft);
  const adjustedTop = unref(isDragging) ? unref(dragY) - unref(boundsTop) : unref(initialTop);
  return {
    left: `${adjustedLeft}px`,
    top: `${adjustedTop}px`,
  };
});
```

**Comment by calebwaldner** [comment]
Unfortunately no, sorry.   
  
  
  
  
  
  
  

  
  
>

**Comment by fardolieri** [comment]
I wrote a helper function that also takes a bounding box as option. It is not very elegant as it does not address the underlying position fixed/absolute offset issue, but it gets the job done for now.

```ts
// imports, for the sake of completeness
import {
  MaybeComputedRef,
  useDraggable,
  UseDraggableOptions,
  useElementBounding,
  useElementSize,
} from '@vueuse/core';
import { Ref, ref, watchEffect } from 'vue';

// setup
const dragElement= ref<HTMLElement>();
const container = ref<HTMLElement>();
const boundingBox = useElementBounding(container );

useDraggableWithBounding(dragElement, { boundingBox }); // <--- use like this


// helper method
function useDraggableWithBounding(
  target: Parameters<typeof useDraggable>[0],
  options: UseDraggableOptions & {
    boundingBox:
      | MaybeComputedRef<{
          left: number;
          right: number;
          top: number;
          bottom: number;
        }>
      | {
          bottom: Ref<number>;
          left: Ref<number>;
          right: Ref<number>;
          top: Ref<number>;
        };
  }
): ReturnType<typeof useDraggable> {
  const boundingBoxRef = ref(options.boundingBox) as Ref<{
    left: number;
    right: number;
    top: number;
    bottom: number;
  }>;

  const drag = useDraggable(target, options);
  const targetRect = useElementSize(target);

  watchEffect(() => {
    drag.position.value.x = Math.max(
      drag.position.value.x,
      boundingBoxRef.value.left
    );
    drag.position.value.x = Math.min(
      drag.position.value.x,
      boundingBoxRef.value.right - targetRect.width.value
    );
    drag.position.value.y = Math.max(
      drag.position.value.y,
      boundingBoxRef.value.top
    );
    drag.position.value.y = Math.min(
      drag.position.value.y,
      boundingBoxRef.value.bottom - targetRect.height.value
    );
  });

  return drag;
}
```

**Comment by flrnt** [comment]
> I have however found a workaround using just vueuse.
> 
> ```ts
> // do this on your container element, if in another vue component save state to store etc.
> const { top: boundsTop, left: boundsLeft } = useElementBounding(container);
> 
> // whatever the initial values are
> const initialLeft = ref(200);
> const initialTop = ref(400);
> 
> // setup use draggable, collect the x,y raw not style
> const { x: dragX, y: dragY, isDragging } = useDraggable(element);
> 
> // generate style manually adjusting left/top
> const style = computed(() => {
>   const adjustedLeft = unref(isDragging) ? unref(dragX) - unref(boundsLeft) : unref(initialLeft);
>   const adjustedTop = unref(isDragging) ? unref(dragY) - unref(boundsTop) : unref(initialTop);
>   return {
>     left: `${adjustedLeft}px`,
>     top: `${adjustedTop}px`,
>   };
> });
> ```

good job. But, in the computed style, once you don't drag it anymore it will return to its initial values

**Comment by laterdays** [comment]
@flrnt  Restrict drag and drop within parent elements only

**Comment by laterdays** [comment]
@flrnt  How do I make useDraggable support absolute drag and limit drag only to the parent element

**Comment by flrnt** [comment]
in the code above, from @calebwaldner the parent element (container) is represented by this line
`const { top: boundsTop, left: boundsLeft } = useElementBounding(container);`
you can use inject/provide to feed the child with parent ref (container). Drop your piece of code if needed

**Comment by calebwaldner** [comment]
I believe @flrnt meant to CC @cmcnicholas correct?
